lib/model/model.go

```diff
@@ -2766,7 +2766,37 @@ func (m *model) BringToFront(folder, file string) {
 
 func (m *model) ResetFolder(folder string) {
 	l.Infof("Cleaning metadata for reset folder %q", folder)
+	m.fmut.RLock()
+	token, ok := m.folderRunnerToken[folder]
+	fcfg := m.folderCfgs[folder]
+	if !ok {
+		// Do it under lock to prevent the folder being started at the same time
+		db.DropFolder(m.db, folder)
+	}
+	m.fmut.RUnlock()
+	if !ok {
+		return
+	}
+
+	// Stop the folder holding restart mutex to prevent any other concurrent
+	// changes from interfering.
+	restartMut := m.folderRestartMuts.Get(folder)
+	restartMut.Lock()
+	defer restartMut.Unlock()
+
+	m.RemoveAndWait(token, 0)
```

Comment:
This seems in principle unsafe, I see no reason that the folder isn't already restarted or reconfigured or removed after we release fmut and before we reach this point. Do we need to allow resetting a running folder at all? It seems to me like we could do all this under fmut, and return an error in case the folder is running...

Reply:
Right, the restart mutes protects from pausing, but one could still entirely remove/readd the folder at the same time. Giving the responsibility to pause the folder to the caller is a good idea :)